# Importing Libraries and Dataframe

In [1]:
# Importing libraries
import pandas as pd
import numpy as np
import os

In [111]:
#Importing 'Products'
path = r'/Users/marialuciaguarita/MLCG Instacart Basket Analysis'
df_prods = pd.read_csv(os.path.join(path, 'Data','Original Data','products.csv'), index_col = False)

# Cleaning data

In [112]:
# Finding missing values from "products.csv"
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [113]:
df_prods.isnull()

product_id  product_name  aisle_id  department_id  prices
0           False         False     False          False   False
1           False         False     False          False   False
2           False         False     False          False   False
3           False         False     False          False   False
4           False         False     False          False   False
...           ...           ...       ...            ...     ...
49688       False         False     False          False   False
49689       False         False     False          False   False
49690       False         False     False          False   False
49691       False         False     False          False   False
49692       False         False     False          False   False

[49693 rows x 5 columns]

In [114]:
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [115]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14   12.20
68             69          NaN        26              7   11.80
115           116          NaN        93              3   10.80
261           262          NaN       110             13   12.10
525           525          NaN       109             11    1.20
1511         1511          NaN        84             16   14.30
1780         1780          NaN       126             11   12.30
2240         2240          NaN        52              1   14.20
2586         2586          NaN       104             13   12.40
3159         3159          NaN       126             11   13.10
3230         3230          NaN       120             16   14.40
3736         3736          NaN        41              8   14.80
4283         4283          NaN        77              7   14.40
4790         4790          NaN        91             16   14.50
38187       38183          NaN        39             12   20.90
40444       40440          NaN       120             16   14.80

In [116]:
#Addressing Missing Values from 'products.csv'
df_prods.shape

(49693, 5)

In [117]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [118]:
df_prods_clean.shape

(49677, 5)

In [119]:
#Finding Duplicated Values
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [120]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11    4.80  
18459        27              5    9.20  
26810        27              5   13.40  
35309       121             14    6.80  
35495       127             11    9.90

In [121]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [122]:
df_prods_clean_no_dups.shape

(49672, 5)

In [123]:
df_prods_clean_no_dups

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19    5.80  
1           104             13    9.30  
2            94              7    4.50  
3            38              1   10.50  
4             5             13    4.30  
...         ...            ...     ...  
49688       124              5    5.30  
49689        42              1    3.10  
49690       112              3    7.80  
49691        41              8    4.70  
49692        73             11   13.50  

[49672 rows x 5 columns]

In [124]:
#Exporting checked data from 'products.csv'
df_prods_clean_no_dups.to_csv(os.path.join(path,'Data','Prepared Data','products_checked.csv'))

In [125]:
pd.set_option("display.float_format",lambda x:"%.2f"% x)
df_prods_clean_no_dups [df_prods_clean_no_dups['prices']==99999.00]

product_id           product_name  aisle_id  department_id   prices
33666       33664  2 % Reduced Fat  Milk        84             16 99999.00

In [126]:
df_prods_clean_no_dups.describe()

product_id  aisle_id  department_id   prices
count    49672.00  49672.00       49672.00 49672.00
mean     24850.35     67.76          11.73     9.99
std      14340.71     38.32           5.85   453.62
min          1.00      1.00           1.00     1.00
25%      12432.75     35.00           7.00     4.10
50%      24850.50     69.00          13.00     7.10
75%      37268.25    100.00          17.00    11.10
max      49688.00    134.00          21.00 99999.00

The maximum value of 99,999.00 is probably wrong.

In [127]:
df_prods_clean_no_dups['prices'].median()

7.1

Probably, the maximum value of prices column  (99,999.00) is wrong. 

In [128]:
#Replacement of the value of 99,999.00 in the "prices" column by the calculated median value
a = 99999.00
b = 7.3
df_prods_clean_no_dups_2 = df_prods_clean_no_dups.replace (a,b)

In [129]:
df_prods_clean_no_dups_2 [df_prods_clean_no_dups_2['product_id']==33664]

product_id           product_name  aisle_id  department_id  prices
33666       33664  2 % Reduced Fat  Milk        84             16    7.30

In [139]:
#Exporting checked data from products
df_prods_clean_no_dups.to_csv(os.path.join(path,'Data','Prepared Data','products_checked1.csv'))

In [153]:
#Importing data from 'orders.csv'
df_ords = pd.read_csv(os.path.join(path, 'Data','Original Data','orders.csv'), index_col = False)
df_ords.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 7 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   order_id                int64  
 1   user_id                 int64  
 2   eval_set                object 
 3   order_number            int64  
 4   order_dow               int64  
 5   order_hour_of_day       int64  
 6   days_since_prior_order  float64
dtypes: float64(1), int64(5), object(1)
memory usage: 182.7+ MB


In [154]:
#Searching for mixed-type data - 'orders.csv'
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:print (col)


No mixed-type data found

In [155]:
#Checking for missing values - 'orders.csv'
df_ords.isnull().sum()

order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [156]:
df_nan=df_ords[df_ords['days_since_prior_order'].isnull()== True]

In [157]:
df_nan

order_id  user_id eval_set  order_number  order_dow  \
0         2539329        1    prior             1          2   
11        2168274        2    prior             1          2   
26        1374495        3    prior             1          1   
39        3343014        4    prior             1          6   
45        2717275        5    prior             1          3   
...           ...      ...      ...           ...        ...   
3420930    969311   206205    prior             1          4   
3420934   3189322   206206    prior             1          3   
3421002   2166133   206207    prior             1          6   
3421019   2227043   206208    prior             1          1   
3421069   3154581   206209    prior             1          3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

Findings: Column 'days_since_prior_order' has 206209 rows missing values. It corresponds to 6% of the data from the mentioned column. Probably many of these values appear because they are from buyers who have never purchased on this platform. What I could do, instead deleting, is create a new column containing the string values "first purchase" and "already a customer". 

In [163]:
df_ords.shape

(3421083, 7)

In [164]:
#Searching for duplicated data - 'orders.csv'
df_dupso = df_ords[df_ords.duplicated()]

In [166]:
df_dupso

Empty DataFrame
Columns: [order_id, user_id, eval_set, order_number, order_dow, order_hour_of_day, days_since_prior_order]
Index: []

In [167]:
df_ords.shape

(3421083, 7)

No duplicated values

In [173]:
#Exporting checked data from 'orders.csv'
df_ords.to_csv(os.path.join(path,'Data','Prepared Data','orders_checked.csv'))